via-router/src/cache.rs

```diff
@@ -34,37 +29,33 @@ impl Cache {
         }
     }
 
-    pub fn read(&self, path: &str) -> Option<Option<Vec<Option<Match>>>> {
-        let lock = &self.entries;
-        let cap = self.capacity;
-
-        #[allow(clippy::never_loop)]
-        let (key, matches) = loop {
-            return match lock.try_read() {
-                Ok(guard) => match fetch(&guard, path) {
-                    Some(existing) => break existing,
-                    None => Some(None),
-                },
-                Err(_) => None,
-            };
-        };
-
-        if cap.checked_div(2).is_some_and(|half| key > half) {
-            if let Ok(mut guard) = lock.try_write() {
+    pub fn promote(&self, key: usize) {
+        if let Ok(mut guard) = self.entries.try_write() {
+            if self.capacity.checked_div(2).is_some_and(|half| key > half) {
                 guard.swap_remove_front(key);
             }
         }
+    }
 
-        Some(Some(matches))
+    pub fn read(&self, path: &str) -> Result<Option<(usize, Matched)>, CacheError> {
+        let guard = self.entries.try_read().or(Err(CacheError))?;
```

Comment:
I typically use if..let for mutex and rw-lock guards but this seems like a reasonable exception.